# Проект по SQL

<b> Заказчик: </b>

- Компания, планирующая покупку крупного сервиса для чтения книг по подписке.

<b> Предоставленные данные: </b>

   - Таблица "books" — cодержит данные о книгах;
   - Таблица "authors" — cодержит данные об авторах;
   - Таблица "publishers" — cодержит данные об издательствах;
   - Таблица "ratings" — cодержит данные о пользовательских оценках книг;
   - Таблица "reviews" — cодержит данные о пользовательских обзорах.

<b> Задача исследования: </b>
- Анализ предоставленных баз данных, с целью формирования ценностного предложения для новго продукта.

In [2]:
from IPython.display import Image
from IPython.core.display import HTML
print('__________________________________________СХЕМА ДАННЫХ__________________________________________')
Image(url= "https://pictures.s3.yandex.net/resources/scheme_1589269096.png")

__________________________________________СХЕМА ДАННЫХ__________________________________________


<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-общей-информации" data-toc-modified-id="Изучение-общей-информации-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение общей информации</a></span></li><li><span><a href="#Решения-заданий" data-toc-modified-id="Решения-заданий-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Решения заданий</a></span><ul class="toc-item"><li><span><a href="#Количество-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-вышедших-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Количество книг, вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-книг" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-книг-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка книг</a></span></li><li><span><a href="#Издательство,-выпустившее-наибольшее-число-книг-толще-50-страниц-(исключение-из-анализа-брошюр)" data-toc-modified-id="Издательство,-выпустившее-наибольшее-число-книг-толще-50-страниц-(исключение-из-анализа-брошюр)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Издательство, выпустившее наибольшее число книг толще 50 страниц (исключение из анализа брошюр)</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг-(книги-с-50-и-более-оценками)" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-(книги-с-50-и-более-оценками)-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг (книги с 50 и более оценками)</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-поставивших-больше-50-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-поставивших-больше-50-оценок-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, поставивших больше 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Изучение общей информации

In [2]:
import pandas as pd
from sqlalchemy import create_engine

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
# Создадим функцию для выполнения запросов
def request(query):
    return pd.io.sql.read_sql(query, con = engine) 

In [5]:
# Список таблиц
table_list = ['books', 'authors', 'publishers', 'ratings', 'reviews']

In [9]:
# В цикле прогоним список таблиц, отобразим любые пять строк и информацию о каждой таблице
for table in table_list:
    print('Таблица "{}"'.format(table))
    display(request('SELECT * FROM {}'.format(table)).sample(5))
    display(request('SELECT * FROM {}'.format(table)).info())
    print('-----------------------------------------------------------------------------------------------------')
    print()

Таблица "books"


book_id  author_id                                              title  \
903      904        565                                     The Waste Land   
135      136        299  Captivating: Unveiling the Mystery of a Woman'...   
768      769        359  The Killing Dance (Anita Blake  Vampire Hunter...   
561      562        512             Ship of Destiny (Liveship Traders  #3)   
401      402        116  Longitude: The True Story of a Lone Genius Who...   

     num_pages publication_date  publisher_id  
903        320       2000-12-01           322  
135        243       2005-04-05           188  
768        368       2006-04-04            47  
561        789       2001-11-27           280  
401        192       2005-10-01           324

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "authors"


author_id                              author
186        187              Gabriel García Márquez
274        275               Jean Craighead George
96          97  Clayton M. Christensen/L.J. Ganser
607        608                 W. Somerset Maugham
371        372                          Lois Lowry

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "publishers"


publisher_id                    publisher
202           203      Oxford University Press
45             46                Berkley Books
314           315        Vintage International
300           301             Turtleback Books
270           271  Simon & Schuster Paperbacks

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "ratings"


rating_id  book_id   username  rating
4520       4521      717     tnovak       5
3839       3840      627   samuel96       3
6088       6089      948   lorichen       4
5970       5971      927     paul88       4
4937       4938      759  jessica49       5

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "reviews"


review_id  book_id         username  \
577         576      211     brandtandrea   
831         832      304           tnovak   
715         714      265     valenciaanne   
1656       1655      600  patriciajohnson   
23           24        9      lesliegibbs   

                                                   text  
577   Per if moment he theory. Its media nearly firs...  
831   Could nearly discover drive where dog policy. ...  
715   Late million lead east leader political. Thoug...  
1656  Rule Democrat each smile possible politics dis...  
23    This station bad edge rather perform. Senior i...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

-----------------------------------------------------------------------------------------------------



Таблицы с данными просмотрены, информация о них изучена. Можно приступать к выполнению заданий.

## Решения заданий

### Количество книг, вышедших после 1 января 2000 года

In [7]:
request(
'''
SELECT 
    COUNT(book_id) AS books_published_after_1_january_2000
FROM 
    books 
WHERE 
    CAST(publication_date AS date) > '2000-01-01'
'''
)

books_published_after_1_january_2000
0                                   819

819 книг было выпущенно после 1 января 2000 года.

### Количество обзоров и средняя оценка книг

In [8]:
request(
'''
SELECT
    books.title AS title,
    subq.count AS cnt_reviews,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
JOIN 
    ratings ON books.book_id = ratings.book_id
JOIN (
    SELECT
        books.book_id,
        books.title,
    COUNT
        (reviews.review_id)
    FROM 
        books
    LEFT JOIN 
        reviews ON books.book_id = reviews.book_id
    GROUP BY 
        books.book_id) AS subq ON books.book_id = subq.book_id
    GROUP BY 
        books.book_id, subq.count
    ORDER BY 
        avg_rating DESC
'''
)

title  cnt_reviews  \
0    The Adventures of Tom Sawyer and Adventures of...            1   
1                      Act of Treason (Mitch Rapp  #9)            2   
2              School's Out—Forever (Maximum Ride  #2)            3   
3                 Stolen (Women of the Otherworld  #2)            2   
4                 Piercing the Darkness (Darkness  #2)            2   
..                                                 ...          ...   
995  The World Is Flat: A Brief History of the Twen...            3   
996                                      Drowning Ruth            3   
997                                              Junky            2   
998                  His Excellency: George Washington            2   
999                               Harvesting the Heart            2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

Количество обзоров и средняя оценка для каждой книги подсчитаны. Интересно, чем так не полюбилась книга "Harvesting the Heart"?

### Издательство, выпустившее наибольшее число книг толще 50 страниц (исключение из анализа брошюр)

In [9]:
request(
'''
SELECT 
    publishers.publisher AS publishing_house,
    COUNT(books.book_id) AS number_of_books
FROM 
    publishers 
    JOIN books ON books.publisher_id = publishers.publisher_id
WHERE 
    books.num_pages > 50
GROUP BY 
    publishers.publisher_id
ORDER BY 
    COUNT(books.book_id) DESC
'''
)

publishing_house  number_of_books
0                   Penguin Books               42
1                         Vintage               31
2        Grand Central Publishing               25
3                Penguin Classics               24
4                Ballantine Books               19
..                            ...              ...
329         Ballantine Books (NY)                1
330                 Plaza y Janés                1
331       HarperCollinsPublishers                1
332  Random House: Modern Library                1
333                 Pan Childrens                1

[334 rows x 2 columns]

Издательством Penguin Books было выпущено 42 книги (объемом больше 50 страниц) и это самый высокий результат среди всех представленных издательств.

### Автор с самой высокой средней оценкой книг (книги с 50 и более оценками)

In [10]:
request(
'''
SELECT 
    authors.author,
    AVG(ratings.rating) AS avg_rating
FROM 
    ratings
    JOIN books ON books.book_id = ratings.book_id
    JOIN authors ON books.author_id = authors.author_id
WHERE 
    ratings.book_id IN (
    SELECT
        ratings.book_id 
    FROM 
        ratings
    GROUP BY 
        ratings.book_id
    HAVING 
        COUNT(ratings.book_id) >= 50
)
GROUP BY 
    authors.author_id
ORDER BY 
    AVG(ratings.rating) DESC
'''
)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.287097
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.246914
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                          Lois Lowry    3.750000
10                                          Dan Brown    3.741259
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

Автор с самой высокой средней оценкой книг - J.K. Rowling (с иллюстрациями Mary GrandPré). Рейтинг знаменитого автора - 4.3

### Среднее количество обзоров от пользователей, поставивших больше 50 оценок

In [11]:
request(
'''
SELECT
    AVG(SUBQ2.review_count) AS avg_review
FROM (
    SELECT 
        reviews.username,
        COUNT(reviews.review_id) AS review_count
    FROM 
        reviews
    RIGHT JOIN (
        SELECT 
            ratings.username,
            COUNT(ratings.rating_id) AS rating_count
        FROM 
            ratings
        GROUP BY 
            ratings.username
        HAVING 
            COUNT(ratings.rating_id) >= 50) AS SUBQ1 ON SUBQ1.username = reviews.username
    GROUP BY reviews.username) AS SUBQ2
'''
)

avg_review
0   24.222222

24 обзора в среднем подготовили пользователи, поставившие больше 50 оценок.

## Вывод

Проведенным исследованием данных, содержащихся в сервисе, установлено:
 - Сервис содержит 1000 книг: 
     - большая часть - 819 книг выпущена после 01.01.2000;
 - Книги написаны 636 авторами: 
     - самый высокооцененный пользователями автор - Джоан Роулинг (4.3 балла);
 - Выпускались собранные книги 340 издательствами: 
     - издательством Penguin Books было выпущено 42 книги (объемом больше 50 страниц) и это самый высокий результат среди всех представленных издательств;
 - 6456 оценок книг проставлено пользователями: 
     - самая низкая оценка у книги Harvesting the Heart (1.5 балла);
 - 2793 обзора подготовлено пользователями 
     - 24 обзора в среднем подготовили пользователи, поставившие больше 50 оценок.